core/src/main/java/jenkins/model/lazy/AbstractLazyLoadRunMap.java

```diff
@@ -353,7 +353,19 @@ public R getByNumber(int n) {
             if (v!=null)        return v;       // already in memory
             // otherwise fall through to load
         }
-        return load(n, null);
+        synchronized (this) {
+            if (index.byNumber.containsKey(n)) { // JENKINS-22767: recheck inside lock
+                BuildReference<R> ref = index.byNumber.get(n);
+                if (ref == null) {
+                    return null;
+                }
+                R v = unwrap(ref);
+                if (v != null) {
+                    return v;
+                }
+            }
```

Comment:
> Seems like it could be moved into that private `Index` class.

Not so easily, since there are three outcomes: `R` (known build), null (known failure), or not yet loaded. Without algebraic types in Java, this would just become more verbose and harder to follow than the existing code. `Optional<Optional<R>>` or `@CheckForNull AtomicReference<R>` are not exactly intuitive.